File: HCrawler.Core/Image.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using HCrawler.Core.Repositories;
using HCrawler.Core.Repositories.Models;

namespace HCrawler.Core
{
    public class Image
    {
        private readonly IImageRepository _imageRepository;

        public Image(IImageRepository imageRepository)
        {
            _imageRepository = imageRepository;
        }

        public Task<IEnumerable<DetailedImage>> GetAllAsync(PageFilter pageFilter)
        {
            return _imageRepository.GetAll(pageFilter);
        }

        public async Task CreateImageIfNotExistsAsync(CreateImage createImage)
        {
            var profileId = await CreateProfileIfNotExistsAsync(createImage);

            var exists = await _imageRepository.ImageExistsAsync(createImage.ImagePath);

            if (!exists)
            {
                await _imageRepository.StoreImageAsync(StoreImage.FromCreateImage(profileId, createImage));
            }
        }

        public async Task<int> CreateProfileIfNotExistsAsync(CreateImage createImage)
        {
            var sourceId = await CreateSourceIfNotExists(createImage);

            var exists = await _imageRepository.ProfileExistsAsync(createImage.ProfileName);

            if (exists)
            {
                return await _imageRepository.GetProfileIdByNameAsync(createImage.ProfileName);
            }

            return await _imageRepository.StoreProfileAsync(StoreProfile.FromCreatImage(sourceId, createImage));
        }

        public async Task<int> CreateSourceIfNotExists(CreateImage createImage)
        {
            var exists = await _imageRepository.SourceExistsAsync(createImage.SourceName);

            if (exists)
            {
                return await _imageRepository.GetSourceIdByNameAsync(createImage.SourceName);
            }

            return await _imageRepository.StoreSourceAsync(StoreSource.FromCreateImage(createImage));
        }
    }
}

File: HCrawler.Core/Image.fs
module HCrawler.Core.Image

open System.Threading.Tasks
open HCrawler.Core.Payloads
open FSharp.Control.Tasks.V2

type Image(imageRepo: IImageRepository) =
    member this.getAllAsync pageFilter = imageRepo.getAllAsync pageFilter

    member this.createSourceIfNotExistsAsync(createImage: CreateImage) =
        task {
            let sourceName = createImage.SourceName

            let! exists = imageRepo.sourceExistsAsync sourceName

            return! match exists with
                    | true -> imageRepo.getSourceIdByNameAsync sourceName
                    | false ->
                        imageRepo.storeSourceAsync
                            { Name = sourceName
                              Url = createImage.SourceUrl }
        }

    member this.createProfileAsync(createImage: CreateImage) =
        task {
            let! sourceId = this.createSourceIfNotExistsAsync createImage

            return! imageRepo.storeProfileAsync
                        { SourceId = sourceId
                          Name = createImage.ProfileName
                          Url = createImage.ProfileUrl }
        }

    member this.createProfileIfNotExistsAsync(createImage: CreateImage) =
        task {
            let profileName = createImage.ProfileName

            let! exists = imageRepo.profileExistsAsync profileName

            return! if exists
                    then imageRepo.getProfileIdByNameAsync profileName
                    else this.createProfileAsync createImage
        }

    member this.createImageAsync(createImage: CreateImage) =
        task {
            let! profileId = this.createProfileIfNotExistsAsync createImage

            return! imageRepo.storeImageAsync
                        { ProfileId = profileId
                          Path = createImage.ImagePath
                          Url = createImage.ImageUrl
                          CreatedOn = createImage.CreatedOn }
        }

    member this.createImageIfNotExistsAsync(createImage: CreateImage) =
        task {
            let imagePath = createImage.ImagePath

            let! exists = imageRepo.imageExistsAsync imagePath

            return! if exists then Task.FromResult(0) else this.createImageAsync createImage
        }
